internal/controllers/machinedeployment/machinedeployment_controller.go

```diff
@@ -238,16 +246,73 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 		if d.Spec.Strategy.RollingUpdate == nil {
 			return ctrl.Result{}, errors.Errorf("missing MachineDeployment settings for strategy type: %s", d.Spec.Strategy.Type)
 		}
-		return ctrl.Result{}, r.rolloutRolling(ctx, d, msList)
+		return ctrl.Result{}, r.rolloutRolling(ctx, d, machineSets)
 	}
 
 	if d.Spec.Strategy.Type == clusterv1.OnDeleteMachineDeploymentStrategyType {
-		return ctrl.Result{}, r.rolloutOnDelete(ctx, d, msList)
+		return ctrl.Result{}, r.rolloutOnDelete(ctx, d, machineSets)
 	}
 
 	return ctrl.Result{}, errors.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
 
+func (r Reconciler) getMachinesForMachineDeployment(ctx context.Context, machineDeployment *clusterv1.MachineDeployment) ([]*clusterv1.Machine, error) {
+	var machineList clusterv1.MachineList
+	if err := r.Client.List(
+		ctx,
+		&machineList,
+		client.MatchingLabels(machineDeployment.Spec.Selector.MatchLabels),
+		client.InNamespace(machineDeployment.GetNamespace()),
+	); err != nil {
+		return nil, err
+	}
+
+	machines := make([]*clusterv1.Machine, 0)
+	for key, _ := range machineList.Items {
+		machines = append(machines, &machineList.Items[key])
+	}
+
+	return machines, nil
+}
+
+func (r Reconciler) syncManagedLabelsInPlace(ctx context.Context, machineSets []*clusterv1.MachineSet, machines []*clusterv1.Machine, desiredNodeLabels *map[string]string) error {
+	// Propagate NodeLabels to MachineSets.
+	for _, machineSet := range machineSets {
+		if machineSet.Spec.Template.Spec.NodeLabels == desiredNodeLabels {
+			continue
+		}
+
+		patchHelper, err := patch.NewHelper(machineSet, r.Client)
+		if err != nil {
+			return err
+		}
+
+		machineSet.Spec.Template.Spec.NodeLabels = desiredNodeLabels
+		if err := patchHelper.Patch(ctx, machineSet); err != nil {
+			return err
+		}
+	}
+
+	// Propagate NodeLabels to Machines.
+	for _, machine := range machines {
+		if machine.Spec.NodeLabels == desiredNodeLabels {
+			continue
+		}
+
+		patchHelper, err := patch.NewHelper(machine, r.Client)
+		if err != nil {
+			return err
+		}
+
+		machine.Spec.NodeLabels = desiredNodeLabels
+		if err := patchHelper.Patch(ctx, machine); err != nil {
+			return err
+		}
+	}
```

Comment:
should this be implemented in the MachineSetController instead, so we have only one controller acting on Machines?